taichi/program/ndarray.cpp

```diff
@@ -103,24 +119,96 @@ std::size_t Ndarray::get_nelement() const {
   return nelement_;
 }
 
+template <typename T>
+T Ndarray::read(const std::vector<int> &I) const {
+  prog_->synchronize();
+  size_t index = flatten_index(total_shape_, I);
+  size_t size_ = sizeof(T);
```

Comment:
How large would T typically be? If it's limited to primitive types, we can reuse the same buffer for these accessors to save the cost of buffer allocation

Reply:
Yup in this case it's typically primitive types, will send a followup PR to cache buffer allocation. :D